onnxruntime/core/session/ort_apis.h

```diff
@@ -375,4 +375,33 @@ ORT_API_STATUS_IMPL(InvokeOp,
 
 ORT_API(void, ReleaseOp, _Frees_ptr_opt_ OrtOp* op);
 
+#ifdef ENABLE_TRAINING_ON_DEVICE
+ORT_API_STATUS_IMPL(CreateTrainingSession,  _In_ const OrtEnv* env, _In_ const OrtSessionOptions* options,
+                  _Outptr_ OrtTrainingSession** out);
+
+ORT_API_STATUS_IMPL(InitializeTrainingSession,  _Inout_ OrtTrainingSession* session, _Inout_ OrtCheckpointState* checkpoint_state,
+                  _In_ const ORTCHAR_T* train_model_path, _In_ const ORTCHAR_T* eval_model_path,
+                  _In_ const ORTCHAR_T* optimizer_model_path);
+
+ORT_API(void, ReleaseTrainingSession, _Frees_ptr_opt_ OrtTrainingSession* session);
+
+ORT_API_STATUS_IMPL(ResetGrad, _Inout_ OrtTrainingSession* session);
+
+ORT_API_STATUS_IMPL(TrainStep, _Inout_ OrtTrainingSession* session, _In_opt_ const OrtRunOptions* run_options,
+                     _In_reads_(input_len) const OrtValue* const* inputs, size_t input_len,
+                     size_t outputs_len, _Inout_updates_all_(outputs_len) OrtValue** outputs);
+
+ORT_API_STATUS_IMPL(EvalStep, _Inout_ OrtTrainingSession* session, _In_opt_ const OrtRunOptions* run_options,
+                     _In_reads_(input_len) const OrtValue* const* inputs, size_t input_len,
+                     size_t outputs_len, _Inout_updates_all_(outputs_len) OrtValue** outputs);
+
+ORT_API_STATUS_IMPL(OptimizerStep, _Inout_ OrtTrainingSession* session, _In_opt_ const OrtRunOptions* run_options);
+
+ORT_API_STATUS_IMPL(LoadCheckpoint, _In_ const ORTCHAR_T* checkpoint_path, _Outptr_ OrtCheckpointState** checkpoint_state);
+
+ORT_API_STATUS_IMPL(SaveCheckpoint, _In_ const ORTCHAR_T* checkpoint_path, _Inout_ OrtTrainingSession* session,
```

Comment:
How do we allow users to provide custom items (like step, epoch...) in the checkpoint state?

Reply:
We need to add more C APIs to allow this... Right now I am marking this as TODO.

Reply:
Another option is adding a key value entry in the same API... we can discuss this later

Reply:
or we can update existing one